builtin/logical/database/backend.go

```diff
@@ -76,13 +76,60 @@ func Backend(conf *logical.BackendConfig) *databaseBackend {
 		Secrets: []*framework.Secret{
 			secretCreds(&b),
 		},
-		Clean:       b.closeAllDBs,
-		Invalidate:  b.invalidate,
-		BackendType: logical.TypeLogical,
+		Clean:             b.closeAllDBs,
+		Invalidate:        b.invalidate,
+		BackendType:       logical.TypeLogical,
+		WALRollbackMinAge: 30 * time.Second,
 	}
 
 	b.logger = conf.Logger
 	b.connections = make(map[string]*dbPluginInstance)
+
+	// Always populate the queue to guard against nil pointers.
+	b.credRotationQueue = queue.NewTimeQueue()
+
+	// But only execute cred rotation if we're the leader.
+	// TODO what if leadership changes during the lifecycle of the application?
+	replicationState := conf.System.ReplicationState()
+	if replicationState.IsLeader() {
+		b.Backend.PeriodicFunc = func(ctx context.Context, req *logical.Request) error {
+			// This will loop until either:
+			// - The queue of passwords needing rotation is completely empty.
+			// - It encounters the first password not yet needing rotation.
+			for _, instance := range b.connections {
+				for {
+					item, err := b.credRotationQueue.PopItem()
+					if err != nil {
+						if err == queue.ErrEmpty {
+							return nil
+						}
+						return err
+					}
+
+					role := item.Value.(*roleEntry)
+					c := dbplugin.StaticUserConfig{
+						Username: role.StaticAccount.Username,
+					}
+
+					if time.Now().Unix() > item.Priority {
+						// We've found our first item not in need of rotation
+						// TODO is this logic correct? Need to also check the logic creating the priority.
+						fmt.Printf("stmt: %+v\n", role.Statements.Rotation)
+						instance.SetCredentials(ctx, c, role.Statements.Rotation)
+						b.credRotationQueue.PushItem(&queue.Item{
+							Key:      item.Key,
+							Value:    role, // TODO is this what needs to be here?
```

Comment:
This is as I intended, yes, however we could forgo a `Value` entirely and load the Role via the `item.Key` too, do you have thoughts?

Reply:
I can see that, and probably prefer the idea of the queue only helping to order events, with no data of its own.